api/plugin.go

```diff
@@ -0,0 +1,602 @@
+	}
+	instance := plugin.GetPluginInstance(userCtx)
+	conf := c.DB.GetPluginConfByUserAndPath(u.ID, plugin.GetPluginInfo().ModulePath)
+	if conf == nil {
+		conf = new(model.PluginConf)
+		conf.UserID = u.ID
+		conf.Enabled = false
+		conf.ModulePath = plugin.GetPluginInfo().ModulePath
+		conf.Token = auth.GeneratePluginToken()
+		switch instance := instance.(type) {
+		case papiv1.Configurer:
+			conf.Config, _ = yaml.Marshal(instance.DefaultConfig())
+		}
+		c.DB.CreatePluginConf(conf)
+	}
+	if err := c.performPluginInstanceInitialization(instance, *conf); err != nil {
+		return err
+	}
+	c.Holder.AddInstance(conf.Token, instance)
+	return nil
+}
+
+func (c *PluginAPI) initPluginForAllUsers(plugin loader.P) error {
+	users := c.DB.GetUsers()
+	for _, u := range users {
+		if err := c.InitPluginForUser(u, plugin); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type pluginStorageHandlerV1 struct {
+	mutex sync.RWMutex
+	db    PluginDatabase
+	conf  model.PluginConf
+}
+
+func (c *pluginStorageHandlerV1) getConf() error {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	conf := c.db.GetPluginConfByToken(c.conf.Token)
+	if conf == nil {
+		return errors.New("configuration not found")
+	}
+	c.conf = *conf
+	return nil
+}
+
+func (c *pluginStorageHandlerV1) updateConf() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.db.UpdatePluginConf(&c.conf)
+}
+
+func (c *pluginStorageHandlerV1) Load() ([]byte, error) {
+	if err := c.getConf(); err != nil {
+		return nil, err
+	}
+
+	return c.conf.Storage, nil
+}
+
+func (c *pluginStorageHandlerV1) Save(d []byte) error {
+	c.conf.Storage = d
+	return c.updateConf()
+}
+
+type pluginMessengerHandlerV1 struct {
+	msgAPI     MessageAPI
+	notifier   Notifier
+	db         PluginDatabase
+	conf       model.PluginConf
+	managedApp *model.Application
+}
+
+func (c *pluginMessengerHandlerV1) initializeApp(ident string) {
+	if c.conf.ApplicationToken != "" {
+		app := c.db.GetApplicationByToken(c.conf.ApplicationToken)
+		if app != nil {
+			// managed application exist
+			c.managedApp = app
+			return
+		}
+	}
+
+	app := &model.Application{
+		Token:       auth.GenerateApplicationToken(),
+		Name:        ident,
+		UserID:      c.conf.UserID,
+		Description: fmt.Sprintf("auto generated application for %s", c.db.GetPluginConfByToken(c.conf.Token).ModulePath),
+	}
+	if err := c.db.CreateApplication(app); err != nil {
+		panic(err)
+	}
+	c.managedApp = app
+	c.conf.ApplicationToken = app.Token
+	c.db.UpdatePluginConf(&c.conf)
+}
+
+func (c *pluginMessengerHandlerV1) SendMessage(msg papiv1.Message) error {
+
+	message := model.Message{
+		ApplicationID: c.managedApp.ID,
+		Message:       msg.Message,
+		Title:         msg.Title,
+		Priority:      msg.Priority,
+		Date:          timeNow(),
+	}
+	c.db.CreateMessage(&message)
+	c.notifier.Notify(c.conf.UserID, &message)
+	return nil
+}
+
+func (c *PluginAPI) performPluginInstanceInitialization(instance loader.Instance, conf model.PluginConf) error {
```

Comment:
Can we extract methods like initPluginForAllUsers, DestroyPluginsForUser, enable/disable to a different struct (inside the plugin package) something like plugin.Manager. I'd like that the api file only contains the swagger docuementation and params parsing (and simple implementations of routes)

Reply:
I have moved the actual initialization code to the loader, but I think it is necessary to leave some methods there(the biggest problem is creating custom webhooks). Moving all initialization code to loader would require routes to be registered in loader, which is also not elegant in my opinion.

Reply:
I dunno too, I'll try to implement what I've on my mind on a different branch. (probably this weekend)

Reply:
Alright, so I'll wait for you then.